cmake/pcl_find_sse.cmake

```diff
@@ -1,210 +1,254 @@
-            else()
-                list(APPEND SSE_FLAGS "-mtune=native")
-            endif()
-            message(STATUS "Using CPU native flags for SSE optimization: ${SSE_FLAGS}")
-        endif()
-    endif()
-
-    # Unfortunately we need to check for SSE to enable "-mfpmath=sse" alongside
-    # "-march=native". The reason for this is that by default, 32bit architectures
-    # tend to use the x87 FPU (which has 80 bit internal precision), thus leading
-    # to different results than 64bit architectures which are using SSE2 (64 bit internal
-    # precision). One solution would be to use "-ffloat-store" on 32bit (see
-    # http://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html), but that slows code down,
-    # so the preferred solution is to try "-mpfmath=sse" first.
-    include(CheckCXXSourceRuns)
-    set(CMAKE_REQUIRED_FLAGS)
+function(PCL_CHECK_FOR_SSE)
+  set(SSE_FLAGS)
+  set(SSE_DEFINITIONS)
 
-    check_cxx_source_runs("
-        // Intel compiler defines an incompatible _mm_malloc signature
-        #if defined(__INTEL_COMPILER)
-            #include <malloc.h>
-        #else
-            #include <mm_malloc.h>
-        #endif
-        int main()
-        {
-          void* mem = _mm_malloc (100, 16);
-          return 0;
-        }"
-        HAVE_MM_MALLOC)
+  if(NOT CMAKE_CROSSCOMPILING)
+    # Test GCC/G++ and CLANG
+    if(CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_GNUCXX OR CMAKE_COMPILER_IS_CLANG)
+      include(CheckCXXCompilerFlag)
+      check_cxx_compiler_flag("-march=native" HAVE_MARCH)
+      if(HAVE_MARCH)
+          list(APPEND SSE_FLAGS "-march=native")
+      else()
+          list(APPEND SSE_FLAGS "-mtune=native")
+      endif()
+      message(STATUS "Using CPU native flags for SSE optimization: ${SSE_FLAGS}")
+    endif()
+  endif()
+
+  # Unfortunately we need to check for SSE to enable "-mfpmath=sse" alongside
+  # "-march=native". The reason for this is that by default, 32bit architectures
+  # tend to use the x87 FPU (which has 80 bit internal precision), thus leading
+  # to different results than 64bit architectures which are using SSE2 (64 bit internal
+  # precision). One solution would be to use "-ffloat-store" on 32bit (see
+  # http://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html), but that slows code down,
+  # so the preferred solution is to try "-mpfmath=sse" first.
+  include(CheckCXXSourceRuns)
+  set(CMAKE_REQUIRED_FLAGS)
+  set(SSE_LEVEL 0)
+
+  check_cxx_source_runs("
+    // Intel compiler defines an incompatible _mm_malloc signature
+    #if defined(__INTEL_COMPILER)
+        #include <malloc.h>
+    #else
+        #include <mm_malloc.h>
+    #endif
+    int main()
+    {
+      void* mem = _mm_malloc (100, 16);
+      return 0;
+    }"
+    HAVE_MM_MALLOC)
+
+  check_cxx_source_runs("
+    #include <stdlib.h>
+    int main()
+    {
+      void* mem;
+      return posix_memalign (&mem, 16, 100);
+    }"
+    HAVE_POSIX_MEMALIGN)
```

Comment:
We should remove this soon. The only place this is used is in linemod (recognition)